app/provider_details/provider_service.py

```diff
@@ -43,35 +44,60 @@ def validate_strategies(self) -> None:
             strategy.validate(notification_type)
 
     def get_provider(self, notification: Notification) -> ProviderDetails:
-        template_or_service_provider_id = self._get_template_or_service_provider_id(notification)
-        if template_or_service_provider_id:
-            provider = get_provider_details_by_id(template_or_service_provider_id)
+        """
+        Return an instance of ProviderDetails.
+        """
+
+        # This is a UUID (ProviderDetails primary key) or None.
+        provider_id: Optional[str] = self._get_template_or_service_provider_id(notification)
```

Comment:
What is the benefit of type checking here? Many providers currently have null for email or sms provider. I believe this will bubble all the way back to the celery task and be caught by the general exception and it will retry this. If we want to type check this I believe we should catch it and throw a `NotificationTechnicalFailureException`.

Reply:
I often add type annotations for variables assigned through a function call because the type is not otherwise obvious.  The type is not enforced.  It's just for readability.  The "Optional" part indicates that it might be None.

Reply:
When I mouse over the called method I see this:
```
(method) _get_template_or_service_provider_id: (notification: Notification) -> (str | None)
Return a primary key for an instance of ProviderDetails using this criteria: 1. Use the notification template's provider_id first. 2. Use the notification service's provider_id if the template's provider_id is null.

The return value, if not None, is a UUID.
```
That's the point of annotations. I think this could be useful for a variable you initialize to `None` but know it will eventually be a `str` (or whatever).